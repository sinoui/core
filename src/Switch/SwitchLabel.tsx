import React, { useState, useCallback } from 'react';
import styled, { css } from 'styled-components';
import classNames from 'classnames';
import { useRipple } from '@sinoui/ripple';
import Color from 'color';
import { Theme } from '@sinoui/theme';
import { getColorFromTheme } from '../utils/color';

export interface Props {
  inputClassName?: string;
  checked?: boolean;
  disabled?: boolean;
  name?: string;
  value?: string;
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
  inputRef?: React.Ref<HTMLInputElement>;
  /**
   * 给input元素应用上新的属性
   */
  inputProps?: React.DetailedHTMLProps<
    React.InputHTMLAttributes<HTMLInputElement>,
    HTMLInputElement
  >;
  color?: string;
}

interface WrapperProps {
  checked?: boolean;
  disabled?: boolean;
  color?: string;
  focused?: boolean;
  theme: Theme;
}

function getColor(props: WrapperProps) {
  return getColorFromTheme(props, props.theme.typography.button.color);
}

export const getHoverBgColor = (props: WrapperProps) => {
  const mainColor = getColor(props);
  if (mainColor && mainColor !== 'inherit' && mainColor !== 'currentColor') {
    return Color(mainColor)
      .alpha(0.12)
      .rgb()
      .string();
  }
  return null;
};

const checkedWrapperStyle = css<WrapperProps>`
  transform: translateX(50%);
  color: ${(props) =>
    getColorFromTheme({
      color: props.color || 'primary',
      theme: props.theme,
      disabled: props.disabled,
    })};

  &:hover {
    background-color: ${(props) =>
      getHoverBgColor({
        color: props.color || 'primary',
        theme: props.theme,
        disabled: props.disabled,
      })};
  }
`;

const disabledWrapperStyle = css<WrapperProps>`
  color: ${(props) =>
    props.theme.palette.type === 'light' ? '#bdbdbd' : '#424242'};

  &:hover {
    background-color: transparent;
  }
`;

const focusedWrapperStyle = css<WrapperProps>`
  background-color: ${(props) =>
    props.checked
      ? getHoverBgColor({
          color: props.color || 'primary',
          theme: props.theme,
          disabled: props.disabled,
        })
      : 'rgba(0, 0, 0, 0.08)'};
`;

const SwitchLabelWrapper = styled.span<WrapperProps>`
  display: flex;
  align-items: inherit;
  justify-content: inherit;
  position: absolute;
  top: 0;
  left: 0;
  color: ${(props) =>
    props.theme.palette.type === 'light' ? ' #fafafa' : '#bdbdbd'};
  z-index: 1;
  transition: ${(props) =>
    props.theme.transitions.create(['left', 'transform'], {
      duration: props.theme.transitions.duration.shortest,
    })};
  padding: 9px;
  border-radius: 50%;

  &:hover {
    background-color: rgba(0, 0, 0, 0.08);
  }

  ${(props) => props.focused && focusedWrapperStyle};
  ${(props) => props.checked && checkedWrapperStyle};
  ${(props) => props.disabled && disabledWrapperStyle};

  & .sinoui-switch__input {
    top: 0;
    left: -100%;
    width: 300%;
    cursor: inherit;
    height: 100%;
    margin: 0;
    opacity: 0;
    padding: 0;
    z-index: 1;
    position: absolute;
  }

  & .sinoui-switch__thumb {
    width: 20px;
    height: 20px;
    box-shadow: 0px 2px 1px -1px rgba(0, 0, 0, 0.2),
      0px 1px 1px 0px rgba(0, 0, 0, 0.14), 0px 1px 3px 0px rgba(0, 0, 0, 0.12);
    border-radius: 50%;
    background-color: currentColor;
  }

  & .sinoui-switch__ripple {
    width: 38px;
    height: 38px;
  }

  & .sinoui-switch__ripple-layout {
    left: 0;
    top: 0;
    width: 38px;
    height: 38px;
  }
`;

/**
 * 开关控件标签组件，主要包括小圆点
 *
 * @export
 * @param {Props} props
 * @returns
 */
export default function SwitchLabel(props: Props) {
  const {
    inputClassName,
    checked,
    disabled,
    color,
    name,
    value,
    onChange,
    inputProps,
    inputRef,
  } = props;

  const [focused, setFocused] = useState(false);

  const onFocus = useCallback(() => setFocused(true), []);

  const onBlur = useCallback(() => setFocused(false), []);

  const handleOnChange = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      if (onChange) {
        onChange(event);
      }
      onBlur();
    },
    [onBlur, onChange],
  );

  const rippleRef = useRipple<HTMLSpanElement>({
    center: true,
    fixSize: true,
    rippleClassName: 'sinoui-switch__ripple',
    rippleLayoutClassName: 'sinoui-switch__ripple-layout',
    disabled,
  });

  return (
    <SwitchLabelWrapper
      ref={rippleRef}
      checked={checked}
      disabled={disabled}
      color={color}
      className={classNames('sinoui-switch__label', {
        'sinoui-switch--focused': focused,
      })}
      focused={focused}
      data-testid="switchLabel"
    >
      <input
        type="checkbox"
        name={name}
        checked={checked}
        onChange={handleOnChange}
        disabled={disabled}
        value={value}
        ref={inputRef}
        onFocus={onFocus}
        onBlur={onBlur}
        {...inputProps}
        className={classNames('sinoui-switch__input', inputClassName)}
      />
      <span className="sinoui-switch__thumb" />
    </SwitchLabelWrapper>
  );
}
