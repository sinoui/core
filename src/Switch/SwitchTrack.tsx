import styled, { css } from 'styled-components';
import { getColorFromTheme } from '../utils/color';

export interface SwitchTrackProps {
  disabled?: boolean;
  checked?: boolean;
  color?: string;
}

const checkedStyle = css<SwitchTrackProps>`
  background-color: ${(props) =>
    getColorFromTheme({
      color: props.color || 'primary',
      theme: props.theme,
      disabled: props.disabled,
    })};
  opacity: 0.5;
`;

const disabledStyle = css<SwitchTrackProps>`
  opacity: ${(props) => (props.checked ? 0.5 : 0.12)};
`;

const SwitchTrack = styled.span<SwitchTrackProps>`
  width: 100%;
  height: 100%;
  opacity: ${(props) => (props.theme.palette.type === 'light' ? 0.38 : 0.3)};
  z-index: -1;
  transition: ${(props) =>
    props.theme.transitions.create(['opacity', 'background-color'], {
      duration: props.theme.transitions.duration.shortest,
    })};
  border-radius: 7px;
  background-color: ${(props) =>
    props.theme.palette.type === 'light'
      ? props.theme.palette.black
      : props.theme.palette.white};

  ${(props) => props.checked && checkedStyle};
  ${(props) => props.disabled && disabledStyle};
`;

export default SwitchTrack;
