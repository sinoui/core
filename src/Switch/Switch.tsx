import React from 'react';
import styled from 'styled-components';
import classNames from 'classnames';
import SwitchTrack from './SwitchTrack';
import SwitchLabel from './SwitchLabel';

interface SwitchProps {
  /**
   * 颜色，默认为主题色
   */
  color?: string;
  /**
   * 不可用状态
   */
  disabled?: boolean;
  /**
   * 选中状态
   */
  checked?: boolean;
  /**
   * 值
   */
  value: string;
  /**
   * 值变更时的回调函数
   */
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
}

const Wrapper = styled.span<{ disabled?: boolean }>`
  width: 58px;
  height: 38px;
  display: inline-flex;
  padding: 12px;
  z-index: 0;
  overflow: hidden;
  position: relative;
  box-sizing: border-box;
  flex-shrink: 0;
  vertical-align: middle;
  cursor: ${(props) => (props.disabled ? 'default' : 'pointer')};
  -webkit-tap-highlight-color: transparent;
`;

/**
 * 开关控件
 *
 * @export
 * @param {SwitchProps} props
 * @returns
 */
export default React.forwardRef<HTMLElement, SwitchProps>(function Switch(
  props,
  ref,
) {
  const { disabled, checked, color } = props;
  return (
    <Wrapper
      className={classNames('sinoui-switch', {
        'sinoui-switch--disabled': disabled,
        'sinoui-switch--checked': checked,
      })}
      disabled={disabled}
      ref={ref}
    >
      <SwitchLabel {...props} />
      <SwitchTrack
        className="sinoui-switch__track"
        disabled={disabled}
        checked={checked}
        color={color}
      />
    </Wrapper>
  );
});
