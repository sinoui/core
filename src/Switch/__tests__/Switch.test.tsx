import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '@sinoui/theme';
import '@testing-library/jest-dom/extend-expect';
import Switch from '../Switch';

describe('Switch 镜像测试', () => {
  it('渲染Switch', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <Switch value="switch1" color="primary" />
        </ThemeProvider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('不可用状态', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <Switch value="switch2" disabled />
        </ThemeProvider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('选中状态', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <Switch value="switch2" checked color="info" />
        </ThemeProvider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
