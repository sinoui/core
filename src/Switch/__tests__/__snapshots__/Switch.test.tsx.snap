// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch 镜像测试 不可用状态 1`] = `
<span
<<<<<<< HEAD
  className="sc-AxiKw iTtuAU sinoui-switch sinoui-switch--disabled"
  disabled={true}
>
  <span
    className="sc-AxirZ cmsVHW sinoui-switch__label"
=======
  className="Switch__Wrapper-m36o27-0 djhusw sinoui-switch sinoui-switch--disabled"
  disabled={true}
>
  <span
    className="SwitchLabel__SwitchLabelWrapper-sc-1o7u2zc-0 cyvRbu sinoui-switch__label"
>>>>>>> origin/bottomNavigation
    data-testid="switchLabel"
    disabled={true}
  >
    <input
      className="sinoui-switch__input"
      disabled={true}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="checkbox"
      value="switch2"
    />
    <span
      className="sinoui-switch__thumb"
    />
  </span>
  <span
<<<<<<< HEAD
    className="sc-AxjAm kAyGwU sinoui-switch__track"
=======
    className="SwitchTrack-sc-1s786x7-0 hXfEpE sinoui-switch__track"
>>>>>>> origin/bottomNavigation
    disabled={true}
  />
</span>
`;

exports[`Switch 镜像测试 渲染Switch 1`] = `
<span
<<<<<<< HEAD
  className="sc-AxiKw dbAppQ sinoui-switch"
>
  <span
    className="sc-AxirZ DKhje sinoui-switch__label"
=======
  className="Switch__Wrapper-m36o27-0 jimTzk sinoui-switch"
>
  <span
    className="SwitchLabel__SwitchLabelWrapper-sc-1o7u2zc-0 gGAMdi sinoui-switch__label"
>>>>>>> origin/bottomNavigation
    color="primary"
    data-testid="switchLabel"
  >
    <input
      className="sinoui-switch__input"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="checkbox"
      value="switch1"
    />
    <span
      className="sinoui-switch__thumb"
    />
  </span>
  <span
<<<<<<< HEAD
    className="sc-AxjAm cMNbLp sinoui-switch__track"
=======
    className="SwitchTrack-sc-1s786x7-0 jazQLR sinoui-switch__track"
>>>>>>> origin/bottomNavigation
    color="primary"
  />
</span>
`;

exports[`Switch 镜像测试 选中状态 1`] = `
<span
<<<<<<< HEAD
  className="sc-AxiKw dbAppQ sinoui-switch sinoui-switch--checked"
>
  <span
    checked={true}
    className="sc-AxirZ iWTLdP sinoui-switch__label"
=======
  className="Switch__Wrapper-m36o27-0 jimTzk sinoui-switch sinoui-switch--checked"
>
  <span
    checked={true}
    className="SwitchLabel__SwitchLabelWrapper-sc-1o7u2zc-0 gZpuOZ sinoui-switch__label"
>>>>>>> origin/bottomNavigation
    color="info"
    data-testid="switchLabel"
  >
    <input
      checked={true}
      className="sinoui-switch__input"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="checkbox"
      value="switch2"
    />
    <span
      className="sinoui-switch__thumb"
    />
  </span>
  <span
    checked={true}
<<<<<<< HEAD
    className="sc-AxjAm glPcoh sinoui-switch__track"
=======
    className="SwitchTrack-sc-1s786x7-0 diCGGz sinoui-switch__track"
>>>>>>> origin/bottomNavigation
    color="info"
  />
</span>
`;
